File: skills/gemini-batch/scripts/get_results.py
#!/usr/bin/env python3
"""Retrieve batch job results.

Usage:
    python get_results.py <job_name>
    python get_results.py batches/abc123 --output results.jsonl

Requirements:
    pip install google-genai
"""

import argparse
import json
import os
import sys

# Load .env file if present
try:
    from dotenv import load_dotenv

    load_dotenv()
except ImportError:
    pass


def get_client():
    """Get Gemini API client."""
    try:
        from google import genai
    except ImportError:
        print("Error: google-genai not installed. Run: pip install google-genai")
        sys.exit(1)

    api_key = os.environ.get("GOOGLE_API_KEY") or os.environ.get("GEMINI_API_KEY")
    if not api_key:
        print("Error: Set GOOGLE_API_KEY or GEMINI_API_KEY environment variable")
        sys.exit(1)

    return genai.Client(api_key=api_key)


def get_results(job_name: str, output: str | None = None) -> str | None:
    """Retrieve batch job results.

    Args:
        job_name: The batch job name/ID
        output: Optional output file path

    Returns:
        Results content or None
    """
    client = get_client()

    batch_job = client.batches.get(name=job_name)
    state = (
        batch_job.state.name
        if hasattr(batch_job.state, "name")
        else str(batch_job.state)
    )

    if state != "JOB_STATE_SUCCEEDED":
        print(f"Warning: Job is not completed. State: {state}")
        return None

    if (
        batch_job.dest
        and hasattr(batch_job.dest, "file_name")
        and batch_job.dest.file_name
    ):
        result_file_name = batch_job.dest.file_name
        print(f"Downloading results from: {result_file_name}")

        file_content = client.files.download(file=result_file_name)
        content = file_content.decode("utf-8")

        if output:
            with open(output, "w", encoding="utf-8") as f:
                f.write(content)
            print(f"Results saved to {output}")
        else:
            print("Results:")
            for line in content.strip().split("\n"):
                try:
                    result = json.loads(line)
                    print(f"  Key: {result.get('key', 'N/A')}")
                    if "response" in result:
                        text = result["response"].get("text", str(result["response"]))
                        print(f"  Response: {text[:200]}...")
                except json.JSONDecodeError:
                    print(f"  {line[:200]}...")

        return content

    elif batch_job.dest and hasattr(batch_job.dest, "inlined_responses"):
        print("Inline Results:")
        for i, resp in enumerate(batch_job.dest.inlined_responses):
            print(f"\nResponse {i + 1}:")
            if hasattr(resp, "response") and resp.response:
                try:
                    print(resp.response.text)
                except AttributeError:
                    print(str(resp.response))
            elif hasattr(resp, "error") and resp.error:
                print(f"Error: {resp.error}")
    else:
        print("No results found.")
        return None


def main():
    parser = argparse.ArgumentParser(description="Retrieve batch job results")
    parser.add_argument("job_name", help="Batch job name/ID")
    parser.add_argument("--output", "-o", help="Output file path for results")

    args = parser.parse_args()

    try:
        get_results(job_name=args.job_name, output=args.output)
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: skills/gemini-batch/scripts/get_results.js
#!/usr/bin/env node
"use strict";

import { writeFile } from "node:fs/promises";
import process from "node:process";
import { parseArgs } from "node:util";

import dotenv from "dotenv";
import { GoogleGenAI } from "@google/genai";

dotenv.config();

function printUsage() {
  console.log(`Retrieve batch job results.

Usage:
  node scripts/get_results.js <job_name>
  node scripts/get_results.js batches/abc123 --output results.jsonl

Options:
  --output, -o    Output file path for result JSONL
  --help, -h      Show this help message

Requirements:
  npm install @google/genai@latest dotenv@latest
`);
}

function getApiKey() {
  const apiKey = process.env.GOOGLE_API_KEY || process.env.GEMINI_API_KEY;
  if (!apiKey) {
    throw new Error("Set GOOGLE_API_KEY or GEMINI_API_KEY environment variable");
  }
  return apiKey;
}

function getState(value) {
  if (!value) return "UNKNOWN";
  if (typeof value === "string") return value;
  if (typeof value.name === "string") return value.name;
  return String(value);
}

async function getResults({ jobName, outputPath }) {
  const ai = new GoogleGenAI({ apiKey: getApiKey() });

  const batchJob = await ai.batches.get({ name: jobName });
  const state = getState(batchJob.state);
  if (state !== "JOB_STATE_SUCCEEDED") {
    console.log(`Warning: Job is not completed. State: ${state}`);
    return null;
  }

  if (batchJob.dest?.fileName) {
    const resultFileName = batchJob.dest.fileName;
    console.log(`Downloading results from: ${resultFileName}`);

    const fileContent = await ai.files.download({ file: resultFileName });
    const content = Buffer.from(fileContent).toString("utf-8");

    if (outputPath) {
      await writeFile(outputPath, content, "utf-8");
      console.log(`Results saved to ${outputPath}`);
    } else {
      console.log("Results:");
      for (const line of content.trim().split("\n")) {
        try {
          const result = JSON.parse(line);
          console.log(`  Key: ${result.key ?? "N/A"}`);
          if (result.response) {
            const text = result.response.text ?? JSON.stringify(result.response);
            console.log(`  Response: ${String(text).slice(0, 200)}...`);
          }
        } catch {
          console.log(`  ${line.slice(0, 200)}...`);
        }
      }
    }
    return content;
  }

  if (batchJob.dest?.inlinedResponses) {
    console.log("Inline Results:");
    batchJob.dest.inlinedResponses.forEach((resp, index) => {
      console.log(`\nResponse ${index + 1}:`);
      if (resp.response) {
        const text = resp.response.text ?? JSON.stringify(resp.response);
        console.log(text);
      } else if (resp.error) {
        console.log(`Error: ${typeof resp.error === "string" ? resp.error : JSON.stringify(resp.error)}`);
      }
    });
    return "inline_results";
  }

  console.log("No results found.");
  return null;
}

function parseCli() {
  const { values, positionals } = parseArgs({
    options: {
      output: { type: "string", short: "o" },
      help: { type: "boolean", short: "h", default: false },
    },
    allowPositionals: true,
  });

  if (values.help) {
    printUsage();
    process.exit(0);
  }

  if (positionals.length === 0) {
    printUsage();
    process.exit(1);
  }

  return {
    jobName: positionals[0],
    outputPath: values.output,
  };
}

async function main() {
  try {
    const options = parseCli();
    await getResults(options);
  } catch (error) {
    console.error(`Error: ${error instanceof Error ? error.message : String(error)}`);
    process.exit(1);
  }
}

await main();
